### PR TITLE
consensus-logic: Remove redundant messaging

### PR DESCRIPTION
**Pull request by bewakes**
## Description
This PR removes the cycle between FCM worker and CSM worker by having CSM send NewState message to FCM via watch channel instead as ForkChoiceMessage::NewState via mpsc channel.

### Type of Change

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature/Enhancement (non-breaking change which adds functionality or enhances an existing one)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update
-   [x] Refactor
-   [ ] New or updated tests

## Checklist

-   [ ] I have performed a self-review of my code.
-   [ ] I have commented my code where necessary.
-   [ ] I have updated the documentation if needed.
-   [ ] My changes do not introduce new warnings.
-   [ ] I have added tests that prove my changes are effective or that my feature works.
-   [ ] New and existing tests pass with my changes.

### CODE REVIEW

File: crates/consensus-logic/src/fork_choice_manager.rs [hunk 1]
```diff
@@ -228,26 +184,27 @@ fn determine_start_tip(
...
-    let init_state = match wait_for_csm_ready(&shutdown, &mut fcm_rx) {
-        Ok(s) => s,
-        Err(e) => {
-            error!(err = %e, "failed to initialize forkchoice manager");
-            return Err(e);
+    info!("waiting for genesis");
+    let init_state = loop {
+        let cstate = status_rx.cl.borrow();
+        if cstate.is_chain_active() && cstate.sync().is_some() {
+            break cstate.clone();
```
[the remark this answers is not in the document]
Comment (bewakes) [reply crates/consensus-logic/src/fork_choice_manager.rs:203]: This lets us get rid of a bunch of code.

File: crates/consensus-logic/src/fork_choice_manager.rs [hunk 2]
```diff
@@ -285,25 +247,35 @@ pub fn tracker_task<D: Database, E: ExecEngineCtl>(
...
-
         let Some(m) = fcm_rx.blocking_recv() else {
             break;
         };
 
+        let csmdata = status_rx.csm.borrow();
+        // Update the status if there's a new state i.e there's a new event
+        if csmdata.last_sync_ev_idx > last_ev_idx {
+            let cstate = status_rx.cl.borrow();
+            let cstate = cstate.deref();
```
[the remark this answers is not in the document]
Comment (bewakes) [reply crates/consensus-logic/src/fork_choice_manager.rs:270]: Cool, will look into `wait_while` and something similar.

> You can just make this *, no?

Probably, will do that if possible.

**Comment by bewakes** [comment]
I had thought about this not being able to act on the updated state immediately. But given how fcm works now, it should be perfectly fine to act on it lazily. Because, what it is doing with the updated states is to update the finalized tip on the tracker. And thus, we can be lazy about that since updating finalized tip is not an expensive task. Also, as I had expected, all the functional tests are passing, so this should be fine(except for the tight loop).

**Comment by delbonis** [comment]
@bewakes So let's say that the chain tip that the FCM thinks is best is on a branch that is now stale since the CSM just finalized a different fork.  Then we get an RPC call requesting the best block, what happens?  We'd return a block that we now know is invalid.

**Comment by bewakes** [comment]
@delbonis, that's a good point that if the rpc is serving the blocks from fcm then we'd be on the stale tip/fork. However, we are serving blocks from the database(L2 provider to be precise) instead of FCM and in future we will have a caching layer for L2 database so fetching blocks from db should not be a problem.

I am not saying being lazy on updating fcm state is the best solution, but this keeps things simpler. Sending new state as a FCM message also seems to be a redundant messaging because we already have `StatusRx` for receiving new state and I think that everything that seeks the latest state should get the data from one place i.e. `StatusRx`.

One other way without having new state as fcm message and without being lazy that I can think of is somehow interleaving the blocking calls(one for fcm message and the other for state update from `StatusRx`) like how `tokio::select` is used. I am not exactly sure about any clever ideas on how this can be done. Do you have any?

>  Are you proposing running the revalidation logic in the RPC handler to account for this so that the incorrect state is never observed?

No, that doesn't sound like a good approach. I just think being lazy might be okay here.

**Comment by delbonis** [comment]
@bewakes 

> However, we are serving blocks from the database(L2 provider to be precise) instead of FCM and in future we will have a caching layer for L2 database so fetching blocks from db should not be a problem.

We use the FCM to determine what the tip is and so which blocks to fetch.  This makes no difference.

> One other way without having new state as fcm message and without being lazy that I can think of is somehow interleaving the blocking calls(one for fcm message and the other for state update from StatusRx) like how tokio::select is used. I am not exactly sure about any clever ideas on how this can be done. Do you have any?

We probably need to rework how the signalling here works instead of just using a channel.  Perhaps something with semaphores?

> No, that doesn't sound like a good approach. I just think being lazy might be okay here.

I presented that hypothetical to illustrate how it's *not* okay.  It also means that we'll have to ensure that the tests still pass when encountering this broken behavior.  That scenario we specifically will need to have a test for.  We'll also have to explain to users in our documentation how occasionally the RPC endpoints just return incorrect results because we didn't feel like implementing it correctly.

We are not longer under the time pressure we were when working towards devnet, we can afford to take some time to implement a *correct* solution.

**Comment by bewakes** [comment]
Apologies for changing the PRs and branches. But closing this PR in favor of https://github.com/alpenlabs/strata/pull/457 as the target has changed from `releases/0.1.0` to `main`.

**Pull request closed without merging**